### ISSUE TITLE
warning C4127: conditional expression is constant in magnum\meshtools\interleave.h(52)

### ISSUE THREAD

**Issue by matjam** [mosra/magnum#238]
A benign warning when compiling with VS:

```
Microsoft (R) Build Engine version 15.6.82.30579 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 3/18/2018 9:37:20 AM.
Project "C:\Users\chrome\Development\games\journey\build\ALL_BUILD.vcxproj" on node 1 (default targets).
Project "C:\Users\chrome\Development\games\journey\build\ALL_BUILD.vcxproj" (1) is building "C:\Users\chrome\Development\games\journey\build\ZERO_CHECK.vcxproj" (2) on node 1 (default targets).
InitializeBuildStatus:
  Creating "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  All outputs are up-to-date.
FinalizeBuildStatus:
  Deleting file "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "C:\Users\chrome\Development\games\journey\build\ZERO_CHECK.vcxproj" (default targets).

Project "C:\Users\chrome\Development\games\journey\build\ALL_BUILD.vcxproj" (1) is building "C:\Users\chrome\Development\games\journey\build\src\JourneyApplication.vcxproj" (3) on node 1 (default targets).
InitializeBuildStatus:
  Touching "JourneyApplication.dir\Debug\JourneyA.1EAB50D7.tlog\unsuccessfulbuild".
CustomBuild:
  All outputs are up-to-date.
VcpkgTripletSelection:
  Not using Vcpkg because VcpkgEnabled is "false"
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\bin\HostX86\x64\CL.exe /c /IC:\Development\libnoise /I"C:\Users\chrome\Development\vcpkg\installed\x64-windows\include\SDL2
  " /I"C:\Users\chrome\Development\vcpkg\installed\x64-windows\include" /I"C:\Users\chrome\Development\vcpkg\installed\x64-windows\include\MagnumExternal\OpenGL" /Zi /nologo /W4 /WX- /diagnostics:classic /Od /Ob0
  /D WIN32 /D _WINDOWS /D _CRT_SECURE_NO_WARNINGS /D _SCL_SECURE_NO_WARNINGS /D NOMINMAX /D WIN32_LEAN_AND_MEAN /D CORRADE_IS_DEBUG_BUILD /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise
  /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"JourneyApplication.dir\Debug\\" /Fd"JourneyApplication.dir\Debug\vc141.pdb" /Gd /TP /wd4251 /wd4244 /wd4267 /wd4351 /wd4373 /wd4510 /wd4610 /wd4512 /wd4661 /wd4702 /wd
  4706 /wd4800 /wd4910 /FC /errorReport:queue C:\Users\chrome\Development\games\journey\src\JourneyApplication.cpp
  JourneyApplication.cpp
c:\users\chrome\development\vcpkg\installed\x64-windows\include\magnum\meshtools\interleave.h(52): warning C4127: conditional expression is constant [C:\Users\chrome\Development\games\journey\build\src\JourneyAppl ication.vcxproj]
  c:\users\chrome\development\vcpkg\installed\x64-windows\include\magnum\meshtools\interleave.h(52): note: consider using 'if constexpr' statement instead
  c:\users\chrome\development\vcpkg\installed\x64-windows\include\magnum\meshtools\interleave.h(52): note: see reference to function template instantiation 'unsigned __int64 Magnum::MeshTools::Implementation::Attr
  ibuteCount::operator ()<std::vector<Magnum::Vector3,std::allocator<_Ty>>,>(const T &) const' being compiled
          with
          [
              _Ty=Magnum::Vector3,
              T=std::vector<Magnum::Vector3,std::allocator<Magnum::Vector3>>
          ]
  c:\users\chrome\development\vcpkg\installed\x64-windows\include\magnum\meshtools\interleave.h(143): note: see reference to function template instantiation 'unsigned __int64 Magnum::MeshTools::Implementation::Att
  ributeCount::operator ()<T,std::vector<Magnum::Vector3,std::allocator<_Ty>>>(const T &,const std::vector<_Ty,std::allocator<_Ty>> &) const' being compiled
          with
          [
              T=std::vector<Magnum::Vector3,std::allocator<Magnum::Vector3>>,
              _Ty=Magnum::Vector3
          ]
  c:\users\chrome\development\games\journey\src\journeyapplication.cpp(45): note: see reference to function template instantiation 'Corrade::Containers::Array<char,void (__cdecl *)(T *,::size_t)> Magnum::MeshTools
  ::interleave<std::vector<Magnum::Vector3,std::allocator<_Ty>>,std::vector<_Ty,std::allocator<_Ty>>>(const std::vector<_Ty,std::allocator<_Ty>> &,const std::vector<_Ty,std::allocator<_Ty>> &)' being compiled
          with
          [
              T=char,
              _Ty=Magnum::Vector3
          ]
Link:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.13.26128\bin\HostX86\x64\link.exe /ERRORREPORT:QUEUE /OUT:"C:\Users\chrome\Development\games\journey\build\src\Debug\JourneyApplicat
  ion.exe" /INCREMENTAL /NOLOGO /LIBPATH:C:/Development/libnoise /LIBPATH:C:/Development/libnoise/Debug "C:\Users\chrome\Development\vcpkg\installed\x64-windows\debug\lib\MagnumSdl2Application-d.lib" "C:\Users\chr
  ome\Development\vcpkg\installed\x64-windows\debug\lib\Magnum-d.lib" "C:\Users\chrome\Development\vcpkg\installed\x64-windows\debug\lib\MagnumMeshTools-d.lib" "C:\Users\chrome\Development\vcpkg\installed\x64-wind
  ows\debug\lib\MagnumPrimitives-d.lib" "C:\Users\chrome\Development\vcpkg\installed\x64-windows\debug\lib\MagnumShaders-d.lib" libnoise.lib "C:\Users\chrome\Development\vcpkg\installed\x64-windows\lib\SDL2.lib" "
  C:\Users\chrome\Development\vcpkg\installed\x64-windows\debug\lib\Magnum-d.lib" "C:\Users\chrome\Development\vcpkg\installed\x64-windows\debug\lib\CorradePluginManager-d.lib" "C:\Users\chrome\Development\vcpkg\i
  nstalled\x64-windows\debug\lib\CorradeUtility-d.lib" opengl32.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"leve
  l='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"C:/Users/chrome/Development/games/journey/build/src/Debug/JourneyApplication.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:/Users
  /chrome/Development/games/journey/build/src/Debug/JourneyApplication.lib" /MACHINE:X64  /machine:x64 JourneyApplication.dir\Debug\JourneyApplication.obj
  JourneyApplication.vcxproj -> C:\Users\chrome\Development\games\journey\build\src\Debug\JourneyApplication.exe
PostBuildEvent:
  setlocal
  powershell -noprofile -executionpolicy Bypass -file C:/Users/chrome/Development/vcpkg/scripts/buildsystems/msbuild/applocal.ps1 -targetBinary C:/Users/chrome/Development/games/journey/build/src/Debug/JourneyAppl
  ication.exe -installedDir C:/Users/chrome/Development/vcpkg/installed/x64-windows/debug/bin -OutVariable out
  if %errorlevel% neq 0 goto :cmEnd
  :cmEnd
  endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone
  :cmErrorLevel
  exit /b %1
  :cmDone
  if %errorlevel% neq 0 goto :VCEnd
  :VCEnd
FinalizeBuildStatus:
  Deleting file "JourneyApplication.dir\Debug\JourneyA.1EAB50D7.tlog\unsuccessfulbuild".
  Touching "JourneyApplication.dir\Debug\JourneyA.1EAB50D7.tlog\JourneyApplication.lastbuildstate".
Done Building Project "C:\Users\chrome\Development\games\journey\build\src\JourneyApplication.vcxproj" (default targets).

InitializeBuildStatus:
  Creating "x64\Debug\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  All outputs are up-to-date.
FinalizeBuildStatus:
  Deleting file "x64\Debug\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild".
  Touching "x64\Debug\ALL_BUILD\ALL_BUILD.tlog\ALL_BUILD.lastbuildstate".
Done Building Project "C:\Users\chrome\Development\games\journey\build\ALL_BUILD.vcxproj" (default targets).
```

The suggested fix doesn't work as it isn't a constant expression.

This seems to shut it up but I don't really understand the best way to solve this.

```
        #pragma warning(push)
        #pragma warning(disable:4127)
        CORRADE_ASSERT(sizeof...(next) == 0 || AttributeCount{}(next...) == first.size() || AttributeCount{}(next...) == ~std::size_t(0), "MeshTools::interleave(): attribute arrays don't have the same length, expected" << first.size() << "but got" << AttributeCount{}(next...), 0);
        #pragma warning(pop)
```

**Comment by mosra** [comment]
Hi, thanks for the report :+1: 

Since this is just an assert, I wouldn't really bother with fixing the warning -- just ignoring it is enough, I guess. The compiler suggests `if constexpr`, but that's C++17, can't use that here yet :) Any other solution would be probably too complex to worth the bother.

I'll integrate your suggestion, thank you :)

**Comment by mosra** [comment]
Fixed in fb483be51275c84e5b4ff964a49270eed4be52c8. I see a bunch of other repetitive warnings while compiling Magnum, will solve them on the go.
